rle.go

```diff
@@ -0,0 +1,1325 @@
+// caller sees in vals as a side effect.
+func newRunContainer32FromVals(alreadySorted bool, vals ...uint32) *runContainer32 {
+	// keep this in sync with newRunContainer32FromArray below
+
+	rc := &runContainer32{}
+	ah := addHelper32{rc: rc}
+
+	if !alreadySorted {
+		sort.Sort(uint32Slice(vals))
+	}
+	n := len(vals)
+	var cur, prev uint32
+	switch {
+	case n == 0:
+		// nothing more
+	case n == 1:
+		ah.m = append(ah.m, interval32{start: vals[0], last: vals[0]})
+		ah.actuallyAdded++
+	default:
+		ah.runstart = vals[0]
+		ah.actuallyAdded++
+		for i := 1; i < n; i++ {
+			prev = vals[i-1]
+			cur = vals[i]
+			ah.add(cur, prev, i)
+		}
+		ah.storeIval(ah.runstart, ah.runlen)
+	}
+	rc.iv = ah.m
+	rc.card = int64(ah.actuallyAdded)
+	return rc
+}
+
+// newRunContainer32FromBitmapContainer makes a new run container from bc.
+func newRunContainer32FromBitmapContainer(bc *bitmapContainer) *runContainer32 {
+
+	rc := &runContainer32{}
+	ah := addHelper32{rc: rc}
+
+	n := bc.getCardinality()
+	it := bc.getShortIterator()
+	var cur, prev, val uint32
+	switch {
+	case n == 0:
+		// nothing more
+	case n == 1:
+		val = uint32(it.next())
+		ah.m = append(ah.m, interval32{start: val, last: val})
+		ah.actuallyAdded++
+	default:
+		prev = uint32(it.next())
+		cur = uint32(it.next())
+		ah.runstart = prev
+		ah.actuallyAdded++
+		for i := 1; i < n; i++ {
+			ah.add(cur, prev, i)
+			if it.hasNext() {
+				prev = cur
+				cur = uint32(it.next())
+			}
+		}
+		ah.storeIval(ah.runstart, ah.runlen)
+	}
+	rc.iv = ah.m
+	rc.card = int64(ah.actuallyAdded)
+	return rc
+}
+
+//
+// newRunContainer32FromArray populates a new
+// runContainer32 from the contents of arr.
+//
+func newRunContainer32FromArray(arr *arrayContainer) *runContainer32 {
+	// keep this in sync with newRunContainer32FromVals above
+
+	rc := &runContainer32{}
+	ah := addHelper32{rc: rc}
+
+	n := arr.getCardinality()
+	var cur, prev uint32
+	switch {
+	case n == 0:
+		// nothing more
+	case n == 1:
+		ah.m = append(ah.m, interval32{start: uint32(arr.content[0]), last: uint32(arr.content[0])})
+		ah.actuallyAdded++
+	default:
+		ah.runstart = uint32(arr.content[0])
+		ah.actuallyAdded++
+		for i := 1; i < n; i++ {
+			prev = uint32(arr.content[i-1])
+			cur = uint32(arr.content[i])
+			ah.add(cur, prev, i)
+		}
+		ah.storeIval(ah.runstart, ah.runlen)
+	}
+	rc.iv = ah.m
+	rc.card = int64(ah.actuallyAdded)
+	return rc
+}
+
+// set adds the integers in vals to the set. Vals
+// must be sorted in increasing order; if not, you should set
+// alreadySorted to false, and we will sort them in place for you.
+// (Be aware of this side effect -- it will affect the callers
+// view of vals).
+//
+// If you have a small number of additions to an already
+// big runContainer32, calling Add() may be faster.
+func (rc *runContainer32) set(alreadySorted bool, vals ...uint32) {
+
+	rc2 := newRunContainer32FromVals(alreadySorted, vals...)
+	p("set: rc2 is %s", rc2)
```

Comment:
rlecommon.go:18
```
// p is a shorthand for fmt.Printf with beginning and                                                                         
// trailing newlines. p() makes it easy                                                                                       
// to add diagnostic print statements.                                                                                        
func p(format string, args ...interface{}) {
    if rleVerbose {
        fmt.Printf("\n"+format+"\n", args...)
    }
}
```
I find godef essential for jumping to definitions. https://github.com/rogpeppe/godef
`(global-set-key [f12] 'godef-jump) ;; in my ~/.emacs`